specifications/ReadersWriters/ReadersWriters.tla

```diff
@@ -0,0 +1,122 @@
+-------------------------- MODULE ReadersWriters --------------------------
+
+EXTENDS FiniteSets, Naturals, Sequences
+
+CONSTANT NumActors
+
+VARIABLES
+    readers, \* set of processes currently reading
+    writers, \* set of processes currently writing
+    waiting  \* queue of processes waiting to access the resource
+
+vars == <<readers, writers, waiting>>
+
+Actors == 1..NumActors
+
+ToSet(s) == { s[i] : i \in DOMAIN s }
+
+fst(s)  == s[1]
+nfst(s) == ~s[1]
+
+WaitingToRead  == { p[2] : p \in ToSet(SelectSeq(waiting, fst)) }
+
+WaitingToWrite == { p[2] : p \in ToSet(SelectSeq(waiting, nfst)) }
+
+---------------------------------------------------------------------------
+
+(***********)
+(* Actions *)
+(***********)
+
+TryRead(actor) ==
+    /\ actor \notin WaitingToRead
+    /\ waiting' = Append(waiting, <<TRUE, actor>>)
```

Comment:
Since writing specifications is more about readability than (space) efficiency, I suggest adding tuples <<"read", actor>> and <<"write", actor>> and adapting the definitions of WaitingToRead and WaitingToWrite accordingly, i.e. something like

PendingReads == { req \in ToSet(waiting) : req[1] = "read" }
WaitingReaders == { req[2] : req \in PendingReads }

which I find easier to read (you may prefer to inline the definition of PendingReads but again, I find it more readable with two definitions). You'd also have to adapt the definition of ReadOrWrite.

Reply:
This is valid. I'll change it.